Source/Particles/NamedComponentParticleContainer.H

```diff
@@ -0,0 +1,178 @@
+    NamedComponentParticleContainer<NewAllocator>
+    make_alike () const {
+        auto tmp = NamedComponentParticleContainer<NewAllocator>(
+            amrex::ParticleContainer<0,0,PIdx::nattribs,0,T_Allocator>::template make_alike<NewAllocator>(),
+            particle_comps,
+            particle_icomps,
+            particle_runtime_comps,
+            particle_runtime_icomps);
+
+        return tmp;
+    }
+
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,T_Allocator>::NumRealComps;
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,T_Allocator>::NumIntComps;
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,T_Allocator>::AddRealComp;
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,T_Allocator>::AddIntComp;
+
+    /** Allocate a new run-time real component
+     *
+     * @param name Name of the new component
+     * @param comm Whether to communicate this component, in the particle Redistribute
+     */
+    void AddRealComp (const std::string& name, bool comm=true)
+    {
+        auto search = particle_comps.find(name);
+        if (search == particle_comps.end()) {
+            particle_comps[name] = NumRealComps();
+            particle_runtime_comps[name] = NumRealComps() - PIdx::nattribs;
+            AddRealComp(comm);
+        } else {
+            amrex::Print() << Utils::TextMsg::Info(
+                name + " already exists in particle_comps, not adding.");
+        }
+    }
+
+    /** Allocate a new run-time integer component
+     *
+     * @param name Name of the new component
+     * @param comm Whether to communicate this component, in the particle Redistribute
+     */
+    void AddIntComp (const std::string& name, bool comm=true)
+    {
+        auto search = particle_icomps.find(name);
+        if (search == particle_icomps.end()) {
+            particle_icomps[name] = NumIntComps();
+            particle_runtime_icomps[name] = NumIntComps() - 0;
+            AddIntComp(comm);
+        } else {
+            amrex::Print() << Utils::TextMsg::Info(
+                name + " already exists in particle_icomps, not adding.");
+        }
+    }
+
+    /** Return the name-to-index map for the compile-time and runtime-time real components */
+    std::map<std::string, int> getParticleComps () const noexcept { return particle_comps;}
+    /** Return the name-to-index map for the compile-time and runtime-time integer components */
+    std::map<std::string, int> getParticleiComps () const noexcept { return particle_icomps;}
+    /** Return the name-to-index map for the runtime-time real components */
+    std::map<std::string, int> getParticleRuntimeComps () const noexcept { return particle_runtime_comps;}
+    /** Return the name-to-index map for the runtime-time integer components */
+    std::map<std::string, int> getParticleRuntimeiComps () const noexcept { return particle_runtime_icomps;}
+
+protected:
+    std::map<std::string, int> particle_comps;
```

Comment:
I think these should all be the faster `std::unordered_map`s, because we don't care about the order of the `key` (`std::string`) values in the container:
https://en.cppreference.com/w/cpp/container/unordered_map